### PR TITLE
Make hibernate optional

### PR DESCRIPTION
**Pull request by jaynel**

### CODE REVIEW

File: src/wpool_queue_manager.erl [hunk 1]
```diff
@@ -30,7 +30,7 @@
 
 -record(state, {wpool                 :: wpool:name(),
                 clients               :: queue(),
-                workers               :: set(),
+                workers               :: gb_sets:set(atom()),
```
Comment (martinsk) [inline src/wpool_queue_manager.erl]: this was supposed to have been changed to a queue

Comment (jaynel) [inline src/wpool_queue_manager.erl]: Separate pull requests. There were too many moving parts and I wanted to make sure we don't have a big bang "we can't merge this now" moment.

File: src/wpool_worker.erl [hunk 2]
```diff
@@ -43,11 +45,24 @@ cast(S, M, F, A) ->
...
 
 %% @private
--spec init(undefined) -> {ok, #state{}}.
-init(undefined) -> {ok, #state{}}.
+-spec init(proplists:proplist()) -> {ok, #state{}} | {ok, #state{}, hibernate | pos_integer()}.
+init(Options) ->
+    case proplists:get_value(hibernate, Options, always) of
+        never     -> {ok, #state{hibernate = never}};
+        always    -> {ok, #state{hibernate = always}, hibernate};
+        when_idle -> {ok, #state{hibernate = when_idle}, ?IDLE_TIMEOUT}
```
Comment (HernanRivasAcosta) [inline src/wpool_worker.erl]: It seems weird that I can't set the idle timeout. Something like:

``` erlang
when_idle -> {ok, #state{hibernate = when_idle}, proplists:get_value(hibernate_timeout, Options, ?IDLE_TIMEOUT)}
```

Would work

File: src/wpool_worker.erl [hunk 3]
```diff
@@ -55,39 +70,49 @@ terminate(_Reason, _State) -> ok.
...
+  reply(State, {error, invalid_request}).
+
+%%% reply/1 and noreply/1,2 are used exclusively so that calls can be traced more easily.
+reply( #state{hibernate = never     } = State, Reply) -> {reply, Reply, State};
+reply( #state{hibernate = always    } = State, Reply) -> {reply, Reply, State, hibernate};
+reply( #state{hibernate = when_idle } = State, Reply) -> {reply, Reply, State, ?IDLE_TIMEOUT}.
+        
+noreply( #state{hibernate = never     } = State) -> {noreply, State};
+noreply( #state{hibernate = always    } = State) -> {noreply, State, hibernate};
+noreply( #state{hibernate = when_idle } = State) -> {noreply, State, ?IDLE_TIMEOUT}.
```
Comment (HernanRivasAcosta) [inline src/wpool_worker.erl]: Same as my previous comment, this probably should not be hardcoded.

**Pull request merged**